Handle unexpected raw values for @objc enums in derived conformances

jrose-apple opened the pull request:
Because people put all sorts of nonsense into `@objc` enums (most reasonably, "private cases", which represent valid values that are not API), the Swift-synthesized implementation of `hash(into:)` needs to not expect a switch statement to be exhaustive. And since Swift-defined `@objc` enums are supposed to behave enough like C-defined enums, they should at least handle simple method calls with an invalid raw value, which means that `rawValue` likewise should not use a switch.

This patch provides alternate implementations that look like this:

```
extension ImportedEnum {
  public var rawValue: Int {
    return unsafeBitCast(self, to: Int.self)
  }

  public func hash(into hasher: inout Hasher) {
    hasher.combine(self.rawValue)
  }
}
```

rdar://problem/41913284

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
@swift-ci Please benchmark

jrose-apple commented:
cc @allevato

jckarter reviewed (approved):
Looks good to me.

slavapestov commented on test/SILGen/enum_raw_representable.swift:
```diff
@@ -1,12 +1,33 @@
 
-// CHECK-DAG: sil [serialized] @$S22enum_raw_representable1EO0B5ValueACSgSi_tcfC
-// CHECK-DAG: sil [serialized] @$S22enum_raw_representable1EO0B5ValueSivg
+@objc public enum CLike: Int {
```
I would prefer this was in its own file, like `objc_enum_raw_representable.swift`

slavapestov commented:
This should be a small code size win too!

itaiferber commented on lib/Sema/DerivedConformanceRawRepresentable.cpp:
```diff
@@ -90,6 +90,27 @@ static void deriveBodyRawRepresentable_raw(AbstractFunctionDecl *toRawDecl) {
   }
 #endif
 
+  if (enumDecl->isObjC()) {
```
This is a good optimization but isn't safe to perform on types which override `rawValue` for some reason, right?

```swift
@objc enum Foo : Int {
    case a
    
    var rawValue: Int {
        return 5
    }
}

let a = Foo.a
print(a.rawValue) // 5
print(unsafeBitCast(a, to: Int.self)) // 0
```

We should probably see if `rawValue` was synthesized, and if so, bypass calling it with this, yeah?

jrose-apple replied:
This is the derived implementation of `rawValue`; it'll only be used if the user didn't write one.

(If you're talking about the `hash(into:)` change, I suppose you're right, but we're doomed either way in that case.)

itaiferber replied:
Sorry, yes! I was looking in the wrong place.

jckarter replied:
`hash(into:)` could use the bitcast approach; isn't that what `rawValue` does?

jrose-apple replied:
I suppose it could. I liked the semantic implications of this one, but bitcasting is a little more correct. Do you think it's important enough to be worth changing?

jckarter replied:
Having hashing use the bitcast value feels more inherently correct to me, since an `@objc` enum can in fact be of any value of its raw type, and a hash implementation ought to feed all of the significant bits of the value to the hasher.

jrose-apple replied:
I realized that `hash(into:)` should probably stay consistent with `==`, which has a default implementation for RawRepresentable types, so I'm going to leave this as is.

jrose-apple commented:
Reorganized tests only.

@swift-ci Please test Linux

jrose-apple commented:
@swift-ci Please smoke test macOS

jrose-apple commented:
@swift-ci Please clean test source compatibility

jrose-apple commented:
@swift-ci Please clean test source compatibility

jrose-apple commented:
(source compat bot is having a problem, so I'm going to hold off for a bit)

jrose-apple commented:
@swift-ci Please smoke test macOS

jrose-apple commented:
@swift-ci Please test Linux

jrose-apple commented:
Unrelated SwiftPM failure?

@swift-ci Please test Linux

jrose-apple committed:
Handle unexpected raw values for @objc enums in derived conformances

Because people put all sorts of nonsense into @objc enums (most
reasonably, "private cases", which represent valid values that are not
API), the Swift-synthesized implementation of 'hash(into:)' needs to
not expect a switch statement to be exhaustive. And since
Swift-defined @objc enums are supposed to behave enough like C-defined
enums, they should at least handle simple method calls with an invalid
raw value, which means that 'rawValue' likewise should not use a
switch.

This patch provides alternate implementations that look like this:

extension ImportedEnum {
  public var rawValue: Int {
    return unsafeBitCast(self, to: Int.self)
  }

  public func hash(into hasher: inout Hasher) {
    hasher.combine(self.rawValue)
  }
}

rdar://problem/41913284

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
(still waiting for the source compat bots to come back online)

jrose-apple commented:
apple/swift-source-compat-suite#213
@swift-ci Please test source compatibility

jrose-apple commented:
Remaining source compat failures match the failures on master. Merging.
